src/Lecture2.hs

```diff
@@ -7,8 +7,316 @@ Stability               : Stable
+    able to continue the fight so they run away.
+
+Implement data types to describe treasure, knight and dragon.
+And implement a function that takes a knight and a dragon and returns
+one of the three possible fight outcomes.
+
+You're free to define any helper functions.
+
+🕯 HINT: If you find the description overwhelming to implement entirely
+  from scratch, try modelling the problem in stages.
+
+    1. Implement all custom data types without using polymorphism.
+    2. Add @newtype@s for safety where you think is appropriate.
+    3. Encode the fight result as a sum type.
+    4. Add polymorphism.
+    5. Make all invalid states unrepresentable. Think, how you can
+       change your types to prevent green dragons from having any
+       treasure besides gold (if you already haven't done this).
+-}
+
+-- some help in the beginning ;)
+data Knight = Knight
+    { knightHealth    :: Int
+    , knightAttack    :: Int
+    , knightEndurance :: Int
+    }
+
+dragonFight = error "TODO"
+
+----------------------------------------------------------------------------
+-- Challenges
+----------------------------------------------------------------------------
+
+{- The following exercises are considered more challenging. However,
+you still may find some of them easier than some of the previous
+ones. Difficulty is a relative concept.
+-}
+
+{- | Write a function that takes a list of numbers and returns 'True'
+if all the numbers are in the increasing order (i.e. the list is
+sorted).
+
+>>> isIncreasing [3, 1, 2]
+False
+>>> isIncreasing [1 .. 10]
+True
+-}
+isIncreasing :: [Int] -> Bool
+isIncreasing [] = True
+isIncreasing [_] = True
+isIncreasing (x : y : xs) = x < y && isIncreasing (y:xs)
+{- | Implement a function that takes two lists, sorted in the
+increasing order, and merges them into new list, also sorted in the
+increasing order.
+
+The lists are guaranteed to be given sorted, so you don't need to
+verify that.
+
+>>> merge [1, 2, 4] [3, 7]
+[1,2,3,4,7]
+-}
+merge :: [Int] -> [Int] -> [Int]
+merge [] [] = []
+merge [] xs = xs
+merge xs [] = xs
+merge (x:xs) (y:ys) = if x < y
+                        then x : merge xs (y : ys)
+                        else y : merge (x : xs) ys
+
+
+{- | Implement the "Merge Sort" algorithm in Haskell. The @mergeSort@
+function takes a list of numbers and returns a new list containing the
+same numbers but in the increasing order.
+
+The algorithm of merge sort is the following:
+
+  1. If the given list has less than 2 elements, it's already sorted.
+  2. Otherwise, split list into two lists of the same size.
+  3. Sort each of two lists recursively.
+  4. Merge two resulting sorted lists to get a new sorted list.
+
+>>> mergeSort [3, 1, 2]
+[1,2,3]
+-}
+mergeSort :: [Int] -> [Int]
+mergeSort = mergeAll . map (:[]) 
+  where
+    mergeAll [] = []
+    mergeAll [t] = t
+    mergeAll xs = mergeAll (mergePairs xs)
+
+    mergePairs (x:y:xs) = merge x y : mergePairs xs
+    mergePairs xs = xs
```

Comment:
Nice and clever bottom-up approach! 🧠 
This implementation also has the desired performance 👍🏻

Reply:
Thank you very much for review !